tmt/export/__init__.py

```diff
@@ -0,0 +1,429 @@
+
+
+# Protocols describing export methods.
+class FmfIdExporter(Protocol):
+    def __call__(self, fmf_id: 'tmt.base.FmfId') -> str:
+        pass
+
+
+class InstanceExporter(Protocol):
+    def __call__(self, instance: ExportableT, keys: Optional[List[str]] = None) -> str:
+        pass
+
+
+class CollectionExporter(Protocol):
+    def __call__(self, collection: List[ExportableT], keys: Optional[List[str]] = None) -> str:
+        pass
+
+
+class Exportable(Generic[ExportableT]):
+    """ Mixin class adding support for exportability of class instances """
+
+    # Declare export plugin registry as a class variable, but do not initialize it. If initialized
+    # here, the mapping would be shared by all classes, which is not a desirable attribute.
+    # Instead, mapping will be created by provides_*_export decorators.
+    _export_plugin_registry: ClassVar[Dict[str, ExportClass]]
+
+    # Keep this method around, to correctly support Python's method resolution order.
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
+        super().__init__(*args, **kwargs)
+
+    @classmethod
+    def provides_export(cls, format: str) -> Callable[[ExportClass], ExportClass]:
+        """
+        A decorator for registering export format.
+
+        Decorate an export plugin class to register a format.
+        """
+
+        def _provides_export(export_cls: ExportClass) -> ExportClass:
+            if not hasattr(cls, '_export_plugin_registry'):
+                cls._export_plugin_registry = {}
+
+            cls._export_plugin_registry[format] = export_cls
+
+            return export_cls
+
+        return _provides_export
+
+    @classmethod
+    def _get_instance_exporter(cls, format: str) -> InstanceExporter:
+        exporter_class = cls._export_plugin_registry.get(format, None)
+
+        if exporter_class is None:
+            raise tmt.utils.GeneralError(
+                f"Export format '{format}' not supported for {cls.__name__.lower()}.")
+
+        return cast(InstanceExporter, getattr(
+            exporter_class, f'export_{cls.__name__.lower()}'))
+
+    @classmethod
+    def _get_collection_exporter(cls, format: str) -> CollectionExporter:
+        exporter_class = cls._export_plugin_registry.get(format, None)
+
+        if exporter_class is None:
+            raise tmt.utils.GeneralError(
+                f"Export format '{format}' not supported for {cls.__name__.lower()}.")
+
+        return cast(CollectionExporter, getattr(
+            exporter_class, f'export_{cls.__name__.lower()}_collection'))
+
+    def _export(self, *, keys: Optional[List[str]] = None) -> _RawExportedInstance:
+        """
+        Export instance as "raw" dictionary.
+
+        The return value is often used by more advanced export methods as their starting
+        position.
+        """
+
+        raise NotImplementedError()
+
+    def export(self, *, format: str, keys: Optional[List[str]] = None, **kwargs: Any) -> str:
```

Comment:
Previously `format_` was used to prevent clash with the builtin function `format`. Are we ok with this override? Also `tmt.base.ExportType` is not used any more. Is this an intentional change?

Reply:
* I'm perfectly fine with that, so far no linter complained, but maybe ruff or pylint will.
* yes, `ExportType` should be removed.

Reply:
`ExportFormat` removed in 472119b885e66532452fea7ca36b5e4a2b2c9ba0

Reply:
Thanks for the `ExportFormat` cleanup. Regarding `format`, if it does not present any issue, let's keep it as it is, without the underscore.

Reply:
If anyone tries to call `format`, they should be stopped by mypy :) And to simplify the playfield, my proposal would be to get rid of `format()` completely, in favor of f-strings, throughout the code base. Then, when noticing `format` in code, one could be sure it's anything but a built-in used for string formatting :)